risc0/zkvm/src/host/server/exec/executor.rs

```diff
@@ -215,7 +217,32 @@ impl<'a> Executor<'a> {
 
     /// Run the executor until [ExitCode::Paused] or [ExitCode::Halted] is
     /// reached, producing a [Session] as a result.
-    pub fn run_with_callback<F>(&mut self, mut callback: F) -> Result<Session>
+    pub fn run_with_callback<F>(&mut self, callback: F) -> Result<Session>
+    where
+        F: FnMut(Segment) -> Result<Box<dyn SegmentRef>>,
+    {
+        let mut guest_session = self.run_guest_only_with_callback(callback)?;
+        match guest_session.exit_code {
+            ExitCode::Fault => {
+                let fault_checker_session = self.run_fault_checker()?;
+                for segment in fault_checker_session.segments {
+                    guest_session.segments.push(segment);
+                }
+                guest_session.journal = fault_checker_session.journal;
```

Comment:
One thing I find uncomfortable about this factoring, of tacking on the fault checker segment and setting the journal, is that it it's clear to me what stops a verifier fr verifying a `Receipt`, getting `Ok(())` back, and then accessing the journal believing the just verified a successful execution. Is there a mechanism to prevent this mistake?